src/Domain/UseCase/UpdateUser.spec.ts

```diff
@@ -0,0 +1,53 @@
+import 'reflect-metadata'
+
+import { User } from '../User/User'
+import { UserRepositoryInterface } from '../User/UserRepositoryInterface'
+import { AuthResponseFactoryInterface } from '../Auth/AuthResponseFactoryInterface'
+import { AuthResponseFactoryResolverInterface } from '../Auth/AuthResponseFactoryResolverInterface'
+
+import { UpdateUser } from './UpdateUser'
+
+describe('UpdateUser', () => {
+  let userRepository: UserRepositoryInterface
+  let authResponseFactoryResolver: AuthResponseFactoryResolverInterface
+  let authResponseFactory: AuthResponseFactoryInterface
+  let user: User
+
+  const createUseCase = () => new UpdateUser(userRepository, authResponseFactoryResolver)
+
+  beforeEach(() => {
+    userRepository = {} as jest.Mocked<UserRepositoryInterface>
+    userRepository.save = jest.fn()
+
+    authResponseFactory = {} as jest.Mocked<AuthResponseFactoryInterface>
+    authResponseFactory.createResponse = jest.fn().mockReturnValue({ foo: 'bar' })
+
+    authResponseFactoryResolver = {} as jest.Mocked<AuthResponseFactoryResolverInterface>
+    authResponseFactoryResolver.resolveAuthResponseFactoryVersion = jest.fn().mockReturnValue(authResponseFactory)
+
+    user = {} as jest.Mocked<User>
+    user.uuid = '123'
+    user.createdAt = new Date(1)
+  })
+
+  it('should update user fields and save it', async () => {
+    expect(await createUseCase().execute({
+      user,
+      updatedWithUserAgent: 'Mozilla',
+      apiVersion: '20190520',
+      version: '004',
+      pwCost: 11,
+      pwSalt: 'qweqwe',
+      pwNonce: undefined
+    })).toEqual({ success: true, authResponse: { foo: 'bar' } })
+
+    expect(userRepository.save).toHaveBeenCalledWith({
```

Comment:
Is this sufficient to ensure that the user model has been updated in the DB? Not really familiar with userRepository.save, but just want to make sure that this would be equivalent to a test where we update user properties, then read the user directly from DB, and ensure DB values match up with the changes we made.

Reply:
`UserRepositoryInterface` is a dependency of this class so we inject a mocked instance of it in the spec.

Ensuring data is saved and/or retrieved to a specific database is an E2E test job on a unit test like these ones. The database is an implementation detail in this architecture (notice we don't know if it's MySQL/Mongo/Redis - we just know the contract in the form of `UserRepositoryInterface`).

The only thing these tests ensure is that proper business logic is executed - methods are called with proper arguments on a given class' dependencies.

If we want to test specific data persisting and/or retrieval - that is ensured IMO in the `SNJS` test suite where it instantiates all the infra (client, syncing-server, db) and makes specific HTTP calls.

Reply:
Hmm interesting..so how do we ensure the logic _within_ the save function has been written correctly?

Would you consider creating a test memory-based store that checks before and after values an E2E test, and if so, when do those tests come into play?

Reply:
Well `save` under the hood in our case is a part of [typeorm](https://typeorm.io/#/) - the `UserRepositoryInterface` is just an abstraction (a wrapper as you will) so that we don't mix the Domain (business logic) with implementation details like SQL/Mongo/Redis queries etc. 

As an addition, infra specific queries are also unit tested in the `MySQLUserRepository.spec.ts` which implements the `UserRepositoryInterface`: https://github.com/standardnotes/syncing-server-js/blob/develop/src/Infra/MySQL/MySQLUserRepository.spec.ts - Though this is missing the `save` spec because that is a part of the typeorm's implementation which the `MySQLUserRepository` extends: https://github.com/typeorm/typeorm/blob/master/src/repository/Repository.ts#L144 so no need for `save` spec in this case. Other methods on the repository are speced as they are our custom ones.

> Would you consider creating a test memory-based store that checks before and after values an E2E test, and if so, when do those tests come into play?
We could do that but then it is redundant to the setup that we have in SNJS. The E2E tests in SNJS already utilize both MySQL as the database and `syncing-server-js` as the syncing server: https://github.com/standardnotes/snjs/blob/master/docker-compose.yml#L65 
Those tests are closer to what we have running in real life then a memory-based implementation of storage because they utilize a MySQL server.